### ISSUE TITLE
[图神经网络市场预测] | GNN Market Prediction

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4441]
**标签**: quantitative-strategy, machine-learning, gnn, market-prediction
### 摘要
使用图神经网络建模市场参与者间的交互关系，捕捉市场微观结构中的协同行为，预测资产价格走势。
### 算法描述
**核心原理**：将市场建模为异构图，节点=资产/机构，边=相关性/持仓关联

**图构建方法**：
1. 资产节点：股票、债券、商品等
2. 机构节点：基金、券商、大户
3. 边类型：收益率相关、持仓重叠、资金流向

**GNN层设计**（GraphSAGE）：
- 消息传递：m_{u→v} = W · h_u
- 聚合邻居：h_v' = σ(W · concat(h_v, AGG({m_{u→v}})))
- 注意力权重：α_{uv} = softmax(e_{uv}) = exp(LeakyReLU(a·[Wh_u||Wh_v])) / Σexp(...)

**价格预测模型**：
P(t+1) = σ(W_out · h_v^T + b_out)

其中 h_v 为资产节点嵌入，W_out 为输出权重矩阵

**特征工程**：
1. 时序特征：收益率、波动率、成交量
2. 图特征：度中心性、PageRank、社区检测
3. 宏观特征：利率、信用利差、VIX

**训练策略**：
- 损失函数：MSE + 图拉普拉斯正则化
- 图拉普拉斯惩罚：λ ||I - D^{-1/2}AD^{-1/2}||_F^2
### Python代码示例
```python
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
import torch.nn.functional as F
from scipy.stats import pearsonr

class GraphAttentionLayer(nn.Module):
    \"\"\"图注意力层\"\"\"
    
    def __init__(self, in_features, out_features, dropout=0.1, alpha=0.2):
        super().__init__()
        self.in_features = in_features
        self.out_features = out_features
        self.dropout = dropout
        self.alpha = alpha
        
        self.W = nn.Parameter(torch.zeros(in_features, out_features))
        self.a = nn.Parameter(torch.zeros(2 * out_features, 1))
        nn.init.xavier_uniform_(self.W)
        nn.init.xavier_uniform_(self.a)
    
    def forward(self, x, adj):
        h = torch.mm(x, self.W)
        N = h.size(0)
        
        a_input = torch.cat([h.repeat(1, N).view(N * N, -1), 
                             h.repeat(N, 1)], dim=1).view(N, -1, 2 * self.out_features)
        e = F.leaky_relu(torch.matmul(a_input, self.a).squeeze(2), self.alpha)
        
        zero_vec = -9e15 * torch.ones_like(e)
        attention = torch.where(adj > 0, e, zero_vec)
        attention = F.softmax(attention, dim=1)
        attention = F.dropout(attention, self.dropout, training=self.training)
        
        h_prime = torch.matmul(attention, h)
        return F.elu(h_prime)

class GNNMarketPredictor(nn.Module):
    \"\"\"基于图神经网络的市场预测模型\"\"\"
    
    def __init__(self, n_features, hidden_dim=64, out_dim=1, n_heads=2):
        super().__init__()
        self.conv1 = GraphAttentionLayer(n_features, hidden_dim, alpha=0.2)
        self.conv2 = GraphAttentionLayer(hidden_dim, hidden_dim, alpha=0.2)
        self.fc = nn.Linear(hidden_dim, out_dim)
        
    def forward(self, x, adj):
        x = F.dropout(x, 0.3, training=self.training)
        x = F.elu(self.conv1(x, adj))
        x = F.dropout(x, 0.3, training=self.training)
        x = F.elu(self.conv2(x, adj))
        x = self.fc(x)
        return x

class MarketGraphBuilder:
    \"\"\"市场图构建器\"\"\"
    
    def __init__(self, correlation_threshold=0.3):
        self.corr_thresh = correlation_threshold
        
    def build_correlation_graph(self, returns_df, min_samples=30):
        \"\"\"基于收益率相关性构建图\"\"\"
        n_assets = returns_df.shape[1]
        
        # 计算相关矩阵
        corr_matrix = returns_df.corr()
        
        # 构建邻接矩阵
        adj = torch.zeros(n_assets, n_assets)
        for i in range(n_assets):
            for j in range(i+1, n_assets):
                corr = corr_matrix.iloc[i, j]
                if abs(corr) > self.corr_thresh:
                    adj[i, j] = adj[j, i] = torch.tensor(corr)
        
        # 添加自环
        adj = adj + torch.eye(n_assets)
        return adj
    
    def build_holding_graph(self, holdings_df, threshold_pct=0.05):
        \"\"\"基于持仓重叠构建图\"\"\"
        n_funds = holdings_df.shape[0]
        n_assets = holdings_df.shape[1]
        
        adj = torch.zeros(n_funds, n_funds)
        for i in range(n_funds):
            for j in range(i+1, n_funds):
                overlap = torch.sum(torch.min(holdings_df[i], holdings_df[j]))
                if overlap > threshold_pct:
                    adj[i, j] = adj[j, i] = overlap
        
        return adj
    
    def compute_graph_features(self, adj):
        \"\"\"计算图统计特征\"\"\"
        degree = adj.sum(dim=1)
        page_rank = self._power_iteration(adj.t(), max_iter=100)
        clustering = self._compute_clustering(adj)
        
        return {
            'degree': degree.numpy(),
            'page_rank': page_rank.numpy(),
            'clustering': clustering.numpy()
        }
    
    def _power_iteration(self, M, max_iter=100, tol=1e-6):
        \"\"\"幂迭代计算PageRank\"\"\"
        n = M.shape[0]
        p = torch.ones(n) / n
        
        for _ in range(max_iter):
            p_new = M @ p
            if torch.norm(p_new - p) < tol:
                break
            p = p_new
        
        return p / p.sum()
    
    def _compute_clustering(self, adj):
        \"\"\"计算聚类系数\"\"\"
        adj_binary = (adj > 0).float()
        n = adj_binary.size(0)
        
        clustering = []
        for i in range(n):
            neighbors = (adj_binary[i] > 0).nonzero(as_tuple=True)[0]
            k = len(neighbors)
            if k < 2:
                clustering.append(0)
            else:
                subgraph = adj_binary[neighbors][:, neighbors]
                edges = subgraph.sum() / 2
                clustering.append(2 * edges.item() / (k * (k - 1)))
        
        return torch.tensor(clustering)

class GNNTradingStrategy:
    \"\"\"GNN策略执行器\"\"\"
    
    def __init__(self, model, graph_builder, lookback=20, forward=5):
        self.model = model
        self.graph_builder = graph_builder
        self.lookback = lookback
        self.forward = forward
        
    def prepare_features(self, price_data, holdings_data=None):
        \"\"\"准备特征矩阵\"\"\"
        returns = price_data.pct_change().dropna()
        
        # 时序特征
        mean_ret = returns.rolling(self.lookback).mean()
        std_ret = returns.rolling(self.lookback).std()
        volume_ratio = returns * 0  # 假设成交量特征
        
        features = torch.tensor(np.column_stack([
            mean_ret.tail(1).values.flatten(),
            std_ret.tail(1).values.flatten(),
            returns.tail(1).values.flatten()
        ]), dtype=torch.float32)
        
        # 构建图
        adj = self.graph_builder.build_correlation_graph(returns)
        
        return features, adj
    
    def predict_returns(self, price_data, holdings_data=None):
        \"\"\"预测未来收益率\"\"\"
        self.model.eval()
        with torch.no_grad():
            features, adj = self.prepare_features(price_data, holdings_data)
            predictions = self.model(features, adj)
        return predictions.numpy().flatten()
    
    def generate_signals(self, price_data):
        \"\"\"生成交易信号\"\"\"
        predictions = self.predict_returns(price_data)
        
        signals = []
        for i, pred in enumerate(predictions):
            if pred > 0.02:
                signals.append({'asset': i, 'action': 'long', 'pred_return': pred})
            elif pred < -0.02:
                signals.append({'asset': i, 'action': 'short', 'pred_return': pred})
            else:
                signals.append({'asset': i, 'action': 'hold', 'pred_return': pred})
        
        return signals
    
    def backtest(self, price_data, initial_capital=1000000):
        \"\"\"回测策略\"\"\"
        capital = initial_capital
        positions = {}
        history = []
        
        for t in range(self.lookback, len(price_data) - self.forward, self.forward):
            window_data = price_data.iloc[t-self.lookback:t]
            signals = self.generate_signals(window_data)
            
            for sig in signals:
                asset = sig['asset']
                action = sig['action']
                
                if action == 'long':
                    positions[asset] = capital * 0.1 / price_data.iloc[t, asset]
                elif action == 'short':
                    positions[asset] = -capital * 0.1 / price_data.iloc[t, asset]
                elif action == 'hold' and asset in positions:
                    del positions[asset]
            
            # 计算收益
            ret = sum(positions.get(i, 0) * (price_data.iloc[t+self.forward, i] / price_data.iloc[t, i] - 1) 
                      for i in positions)
            capital *= (1 + ret)
            
            history.append({'t': t, 'capital': capital, 'positions': dict(positions)})
        
        return history

# 使用示例
np.random.seed(42)
n_days = 300
n_assets = 50

# 模拟价格数据
price_data = pd.DataFrame(
    np.exp(np.cumsum(np.random.randn(n_days, n_assets) * 0.02, axis=0)),
    columns=[f'Asset_{i}' for i in range(n_assets)]
)

# 构建市场图
graph_builder = MarketGraphBuilder(correlation_threshold=0.4)

# 训练模型
model = GNNMarketPredictor(n_features=3, hidden_dim=32, out_dim=1)
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

print(\"模型初始化完成，开始训练...\")
print(f\"资产数量: {n_assets}\")
print(f\"训练天数: {n_days}\")

strategy = GNNTradingStrategy(model, graph_builder, lookback=20, forward=5)
signals = strategy.generate_signals(price_data)
print(f\"生成信号数: {len(signals)}\")
```
### 数据来源
- Reddit: r/quant, r/algorithmictrading, r/Finance
- 小红书: 量化投资、理财技巧
- 知乎: 量化交易、机器学习交易
- 其他: Investopedia, Quantopian, arXiv
